OAK-9447: Support for Mongo Java Driver 4.8

Joscorbe opened the pull request.

Joscorbe committed:
Support Mongo Driver 4.2:
- Removed some deprecated methods.
- Removed Socket Keep Alive, as it has been removed from Mongo in 4.0.
- Added ClusterListener to get the description of the server.
- Removed usage getReplicaSetStatus() as it disappears in Mongo 4.0.
- Removed tests of deprecated features (like keep alive).

Joscorbe committed:
Added ClusterListener to default Builder.

Joscorbe committed:
* Support Mongo Driver 4.2:
- Moved MONGODB_PREFIX to MongoConnection, as MongoURI class is deprecated and removed.
- Removed usage of FSYNC_SAFE, JOURNAL_SAFE, REPLICAS_SAFE and SAFE as doesn't exist anymore.
- Changed name of ReplicaSetStatusListener to OakClusterListener, as it is now more generic.

Joscorbe committed:
Upgraded mongo driver to version 4.2.1.
Removed deprecated and non existent methods in the new version.

Joscorbe committed:
Disabled call to removed "modifiers" method.

Joscorbe committed:
Refactor MongoClusterListener to a new class.

Joscorbe committed:
Refactor methods for new Listener to MongoConnection class.

Joscorbe committed:
Add defaultBuilder to DocumentMongoFixture

Joscorbe committed:
Reverted these disabled tests.

Joscorbe committed:
Removed usage of deprecated ServerMonitorListenerAdapter.

Joscorbe committed:
Implemented initialization check for replicaSetStatus using a Latch.

Joscorbe committed:
Added missing license to MongoClusterListener class.

mreutegg commented on oak-parent/pom.xml:
```diff
@@ -54,7 +54,7 @@
     <lucene.version>4.7.1</lucene.version>
     <solr.version>8.6.3</solr.version>
-    <mongo.driver.version>3.12.7</mongo.driver.version>
+    <mongo.driver.version>4.2.1</mongo.driver.version>
```
Can you please update to 4.2.3? That seems to be the most recently available.

Joscorbe replied:
Yes, upgraded to 4.2.3

mreutegg commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java:
```diff
@@ -300,4 +308,65 @@ public static ReadConcernLevel readConcernLevel(ReadConcern readConcern) {
+        if (listener != null) {
+            return listener.isReplicaSet();
+        }
+        System.out.println("Method isReplicaSet called for a MongoClient without any OakClusterListener!");
```
Please do not use System.out. Writing to the logger is sufficient.

mreutegg commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java:
```diff
@@ -300,4 +308,65 @@ public static ReadConcernLevel readConcernLevel(ReadConcern readConcern) {
+        if (listener != null) {
+            return listener.getServerAddress();
+        }
+        System.out.println("Method getAddress called for a MongoClient without any OakClusterListener!");
```
Same as above.

mreutegg commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java:
```diff
@@ -300,4 +308,65 @@ public static ReadConcernLevel readConcernLevel(ReadConcern readConcern) {
+                return replClusterListener;
+            }
+        }
+        System.out.println("No OakClusterListener found for this MongoClient!");
```
Same as above.

mreutegg commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java:
```diff
@@ -300,4 +308,65 @@ public static ReadConcernLevel readConcernLevel(ReadConcern readConcern) {
+            }
+        }
+        System.out.println("No OakClusterListener found for this MongoClient!");
+        LOG.warn("No OakClusterListener found for this MongoClient!");
```
In this case, I wouldn't log a warning. Returning null when there is no OakClusterListener is valid. I would leave it up to the caller to decide whether a WARN message is appropriate, depending on the context.

mreutegg commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java:
```diff
@@ -27,18 +27,27 @@
 
+    public static final String MONGODB_PREFIX = "mongodb://";
+
+    private static final Logger LOG = LoggerFactory.getLogger(MongoUtils.class);
```
This is the wrong class for the logger. It should get the logger for `MongoConnection`.

mreutegg commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoUtils.java:
```diff
@@ -31,18 +29,19 @@
  * Provides static utility methods for MongoDB.
  */
-class MongoUtils {
+public class MongoUtils {
```
This is unnecessary because its only incorrect use is to get a logger for this class in `MongoConnection`.

Joscorbe replied:
After re-checking this I think I have to keep the "public", as there is a test in a different package that needs to access the MongoUtils anyway.

mreutegg replied:
I'm still not able to identify those tests. I'm able to build `oak-store-document` module successfully when I revert the changes on this class, that is, making it package private again.

mreutegg commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDocumentNodeStoreBuilderBase.java:
```diff
@@ -116,25 +114,6 @@ public T setMongoDB(@NotNull MongoClient client,
-     * @param enable whether to enable or disable it.
-     * @return this
-     */
-    public T setSocketKeepAlive(boolean enable) {
```
To ease upgrades I'd suggest deprecating the methods first and then remove it in a subsequent Oak release.

mreutegg commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoClusterListener.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    @Override
+    public void clusterOpening(ClusterOpeningEvent event) {
```
The default implementation is already a NOP. Overriding the method is not necessary.

mreutegg commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoClusterListener.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    @Override
+    public void clusterClosed(ClusterClosedEvent event) {
```
Same as above.

mreutegg commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoClusterListener.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    public ServerAddress getServerAddress() {
+        try {
```
Can you please move the latch.await and exception handling into a separate method and reuse?

mreutegg commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoClusterListener.java:
```diff
@@ -0,0 +1,88 @@
+        latch = new CountDownLatch(1);
+    }
+
+    public ServerAddress getServerAddress() {
```
Can you please add a `@Nullable` annotation for the return?

mreutegg commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoClusterListener.java:
```diff
@@ -0,0 +1,88 @@
+        return serverAddress;
+    }
+
+    public ServerAddress getPrimaryAddress() {
```
Can you please add a `@Nullable` annotation for the return?

mreutegg commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MongoDbTest.java:
```diff
@@ -201,7 +201,8 @@ public void updateDocument() {
                                          Bson query) {
-        return collection.find(query).modifiers(new BasicDBObject("$explain", true)).first();
+        //TODO: Determine if this test is still needed. Modifiers is removed in driver 4, so this has to be reworked
+        return collection.find(query)/*.modifiers(new BasicDBObject("$explain", true))*/.first();
```
While those tests are all ignored, this can still be fixed. I think simply calling `collection.find(query).explain(BasicDBObject.class)` will do the trick.

mreutegg commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoTestClient.java:
```diff
@@ -19,9 +19,11 @@
 import java.util.concurrent.atomic.AtomicReference;
 
 import com.mongodb.MongoClient;
+import com.mongodb.MongoClientOptions;
```
This import is not used.

mreutegg commented:
@Joscorbe, this PR currently has conflicts. Can you please rebase your changes and resolve the conflicts?

Joscorbe committed:
Removed deprecated option socketKeepAlive and fixed some errors during the rebase.

Joscorbe committed:
Applied proposals by Marcel during review.

mreutegg commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java:
```diff
@@ -27,18 +27,27 @@
 import static com.google.common.base.Preconditions.checkNotNull;
 
+import com.mongodb.event.ClusterListener;
+import org.apache.jackrabbit.oak.plugins.document.mongo.MongoUtils;
```
This import is unused and can be removed.
```suggestion
```

mreutegg commented:
I think the import package in `oak-store-document/pom.xml` must be adjusted as well. It currently limits the version scope to something less than 4.x.

Joscorbe committed:
Removed unused MongoUtils imports.

Joscorbe committed:
Adjusted the import package to allow 4.x versions.

Joscorbe committed:
Removed public for MongoUtils class.

mreutegg commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoUtils.java:
```diff
@@ -31,12 +29,13 @@
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
+import java.util.Set;
```
Can you please revert the changes on this file entirely? There's only rearrangement of import statements left.

Joscorbe committed:
Cleaned some imports

Joscorbe committed:
Upgraded mongodb-legacy to version 4.7.1
Use Mongo server 4.2 by default, instead of 3.6.

mreutegg commented:
@Joscorbe it seems to me only a few changes would be necessary to avoid use of the legacy driver, but I may be missing some less trivial changes. https://www.mongodb.com/docs/drivers/java/sync/current/legacy/#api-changes

WDYT?

Joscorbe commented:
@mreutegg Yes, legacy driver is the one I'm using for these changes, you can see it here: 
https://github.com/apache/jackrabbit-oak/pull/295/files#diff-61d831053d713cff9598e416a9aafe8a69a1ee87c8bec8226d805f388e2173f8

All of the changes I had to apply are mainly because of the breaking changes listed here since version 3.12:
https://www.mongodb.com/docs/drivers/java/sync/current/upgrade/

We were still using some deprecated methods that were removed, so I had to change them even to be able to build the project.

Joscorbe commented:
Hi @mreutegg, yes, the plan is to move to sync driver after the failing tests in this build are fixed. Apart from the documentation you have showed, I have this support case where they add some points about the transition from legacy to sync:

https://www.mongodb.com/community/forums/t/conversion-from-mongodb-driver-legacy-to-mongodb-driver-sync/125411

Joscorbe committed:
Improved testMongoAvailability() for oak-upgrade tests.

Joscorbe committed:
Merge branch 'trunk' into mongo-4.2

Joscorbe committed:
Merge remote-tracking branch 'oak/trunk' into mongo-4.2

Joscorbe commented:
All checks have passed at this point, I will continue now the migration to sync driver in this same PR.

Joscorbe committed:
Migrate from legacy to sync Mongo Java Driver

Joscorbe committed:
Merge branch 'trunk' into mongo-4.2

Joscorbe commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java:
```diff
@@ -305,13 +303,12 @@ private void registerNodeStore() throws IOException {
                         mongoURI.getHosts(), db, config.cache(), persistentCache,
                         journalCache, config.blobCacheSize(), config.maxReplicationLagInSecs());
-                log.info("Mongo Connection details {}", MongoConnection.toString(mongoURI.getOptions()));
+                log.info("Mongo Connection details {}", MongoConnection.toString(mongoURI));
```
I have added a new method MongoConnection.toString that accepts a ConnectionString as a parameter.

Joscorbe commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java:
```diff
@@ -142,41 +153,53 @@ public void close() {
-                .getOptions().getWriteConcern();
-        return !WC_UNKNOWN.equals(wc);
+        WriteConcern wc = new ConnectionString(checkNotNull(uri)).getWriteConcern();
+        return wc != null && !WC_UNKNOWN.equals(wc);
```
WriteConcern can be null now

Joscorbe commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java:
```diff
@@ -186,9 +209,8 @@ public static boolean hasWriteConcern(@NotNull String uri) {
-                .getOptions().getReadConcern();
-        return readConcernLevel(rc) != null;
+        ReadConcern rc = new ConnectionString(checkNotNull(uri)).getReadConcern();
+        return rc != null && readConcernLevel(rc) != null;
```
ReadConcern can be null now

Joscorbe commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoTestClient.java:
```diff
@@ -16,31 +16,136 @@
 import org.jetbrains.annotations.NotNull;
 
-class MongoTestClient extends MongoClient {
+class MongoTestClient implements MongoClient {
```
This is now an interface and I have found no better way to achieve the same result.

mreutegg replied:
The only alternative that comes to my mind is using `java.lang.reflect.Proxy`, but implementing the methods is also fine with me.

Joscorbe commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/mongo/ReadPreferenceIT.java:
```diff
@@ -116,10 +116,11 @@ public void testReadWriteMode() {
 
-        mongoDS.setReadWriteMode("readPreference=secondary&w=2&safe=true&j=true");
+        // Command failed with error 2 (BadValue): 'cannot use 'w' > 1 when a host is not replicated'
+        mongoDS.setReadWriteMode("readPreference=secondary&w=1&safe=true&j=true");
```
I think this change doesn't affect in what the test wants to check.
Unless we start the cluster as a replica set, Mongo doesn't let you specify a 'w' value bigger than 1.

Joscorbe commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/mongo/ReadPreferenceIT.java:
```diff
@@ -116,10 +116,11 @@ public void testReadWriteMode() {
 
         assertEquals(ReadPreference.secondary(), mongoDS.getDBCollection(NODES).getReadPreference());
-        assertEquals(2, mongoDS.getDBCollection(NODES).getWriteConcern().getW());
+        assertEquals(1, mongoDS.getDBCollection(NODES).getWriteConcern().getW());
```
Same as before

Joscorbe committed:
Remove wrong line

mreutegg commented on oak-store-document/pom.xml:
```diff
@@ -44,8 +44,8 @@
-              com.mongodb*;version="[3.8, 4)";resolution:=optional,
-              org.bson*;version="[3.8, 4)";resolution:=optional,
+              com.mongodb*;version="[3.8, 5)";resolution:=optional,
+              org.bson*;version="[3.8, 5)";resolution:=optional,
```
I think the lower bound needs adjustment as well, or does oak-store-document with the changes in this PR still work with version 3.8 of the MongoDB Java driver?

Joscorbe replied:
No, they aren't compatible. I have upgraded them in this file. The minimum MongoDB Driver version is now 4.7. I thought it could be a lower one like 4.0, but due to some breaking changes introduced in several classes they are not compatible.

mreutegg commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDocumentNodeStoreBuilderBase.java:
```diff
@@ -104,18 +103,11 @@ public T setMongoDB(@NotNull MongoClient client,
      * @param enable whether to enable or disable it.
```
Can you please update JavaDoc and mention this is deprecated and calling this method does not have any effect? Socket keep-alive is always enabled.

mreutegg commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java:
```diff
@@ -142,41 +153,53 @@ public void close() {
+    public static MongoClientSettings.Builder getDefaultBuilder() {
+        return MongoClientSettings.builder()
+                .applicationName("Oak DocumentMK")
+                .applyToConnectionPoolSettings(builder -> {
```
Curly brackets are unnecessary here. Same for the two other that follow.

mreutegg commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceTest.java:
```diff
@@ -203,42 +189,6 @@ public void persistentCacheExclude() throws Exception{
         assertFalse(dns.getNodeCachePredicate().apply(Path.fromString("/x")));
     }
 
-    @Test
-    public void preset() throws Exception {
-        MockOsgi.setConfigForPid(context.bundleContext(),
-                Configuration.PRESET_PID,
-                DocumentNodeStoreServiceConfiguration.PROP_SO_KEEP_ALIVE, true);
-        MockOsgi.activate(preset, context.bundleContext());
-
-        MockOsgi.setConfigForPid(context.bundleContext(), PID, newConfig(repoHome));
-        MockOsgi.activate(service, context.bundleContext());
-
-        DocumentNodeStore store = context.getService(DocumentNodeStore.class);
-        MongoDocumentStore mds = getMongoDocumentStore(store);
-        assertNotNull(mds);
-        MongoClient client = MongoDocumentStoreTestHelper.getClient(mds);
-        assertTrue(client.getMongoClientOptions().isSocketKeepAlive());
-    }
-
-    @Test
-    public void presetOverride() throws Exception {
-        MockOsgi.setConfigForPid(context.bundleContext(),
-                Configuration.PRESET_PID,
-                DocumentNodeStoreServiceConfiguration.PROP_SO_KEEP_ALIVE, true);
-        MockOsgi.activate(preset, context.bundleContext());
-
-        Map<String, Object> config = newConfig(repoHome);
-        config.put(DocumentNodeStoreServiceConfiguration.PROP_SO_KEEP_ALIVE, false);
-        MockOsgi.setConfigForPid(context.bundleContext(), PID, config);
-
-        MockOsgi.activate(service, context.bundleContext());
-
-        DocumentNodeStore store = context.getService(DocumentNodeStore.class);
-        MongoDocumentStore mds = getMongoDocumentStore(store);
-        MongoClient client = MongoDocumentStoreTestHelper.getClient(mds);
-        assertFalse(client.getMongoClientOptions().isSocketKeepAlive());
-    }
-
```
Can you please re-introduce these tests again and use another configuration that is still available to exercise the configuration preset feature?

Joscorbe replied:
Yes, I have re-introduced both tests using the property PROP_UPDATE_LIMIT

mreutegg commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongodProcessFactory.java:
```diff
@@ -145,7 +147,8 @@ private void initRS(String rs, int[] ports) {
         Document config = new Document("_id", rs);
         config.append("members", members);
-        try (MongoClient c = new MongoClient(localhost(), ports[0])) {
+        ConnectionString mongoURI = new ConnectionString(localhost());
+        try (MongoClient c = MongoClients.create(localhost())) {
```
This does not look right to me. `mongoURI` is never used and IIUC the MongoClient created here will try to connect to a MongoDB instance on the default port, which isn't necessarily the case.

mreutegg commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDocumentStore.java:
```diff
@@ -1924,15 +1924,15 @@ public void setReadWriteMode(String readWriteMode) {
-            MongoClientURI uri = new MongoClientURI(rwModeUri);
-            ReadPreference readPref = uri.getOptions().getReadPreference();
+            ConnectionString uri = new ConnectionString(rwModeUri);
+            ReadPreference readPref = uri.getReadPreference();
```
As you noted in another comment `ReadPreference` as well as `WriteConcern` further down can be `null` after the driver update. There need to be null checks.

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoClusterListener.java:
```diff
@@ -0,0 +1,82 @@
+    // didn't have time to receive the description from the cluster. This latch is used to check
+    // if the connection was properly initialized.
+    private final CountDownLatch LATCH;
+    private final int LATCH_AWAIT_TIMEOUT = 15;
```
This could be a compile time constant.

rishabhdaim replied:
```suggestion
    private static final int LATCH_AWAIT_TIMEOUT = 15;
```

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoClusterListener.java:
```diff
@@ -0,0 +1,82 @@
+    // Sometimes we need to wait a few seconds in case the connection was just created, the listener
+    // didn't have time to receive the description from the cluster. This latch is used to check
+    // if the connection was properly initialized.
+    private final CountDownLatch LATCH;
```
```suggestion
    private final CountDownLatch latch;
```
Only compile time constants should be in upper case.

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoClusterListener.java:
```diff
@@ -0,0 +1,82 @@
+
+    @Override
+    public void clusterDescriptionChanged(final ClusterDescriptionChangedEvent event) {
+        for (ServerDescription sd : event.getNewDescription().getServerDescriptions()) {
```
```suggestion
        Optional<ServerDescription> ol = Iterables.tryFind(event.getNewDescription().getServerDescriptions(), s -> s.getState() == ServerConnectionState.CONNECTED);
        if (ol.isPresent()) {
            ServerDescription sd = ol.get();
            serverAddress = sd.getAddress();
            primaryAddress = new ServerAddress(sd.getPrimary());
            if (sd.isReplicaSetMember()) {
                // Can't assign directly the result of the function because in some cases the cluster
                // type is UNKNOWN, mainly when the cluster is changing it's PRIMARY.
                replicaSet = true;
            }
            latch.countDown();
        } else {
            // log here that no connected state is present
        }
```

Joscorbe replied:
I have applied your recommendation. However, I don't think we should add a log message when no connected state is present, as this event can be executed in multiple situations (for example, when there is a disconnection), meaning we could flood the logs with messages that are not useful.

What I'm adding now is a single log message when `getPrimaryAddress()`, `getAddress()` or `isReplicaSet()` methods return a null value because the connection was not properly established. This way we ensure the log message will show up just once.

Joscorbe replied:
I have changed the first part of the code to this one, I think it's equivalent and doesn't have to use the Google libraries for Optional and Iterables, but instead it uses Java API code:
`Optional<ServerDescription> ol = event.getNewDescription().getServerDescriptions()
            .stream().filter(s -> s.getState() == ServerConnectionState.CONNECTED).findFirst();`

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoClusterListener.java:
```diff
@@ -0,0 +1,82 @@
+
+    @Nullable
+    public ServerAddress getServerAddress() {
+        waitForLatch();
```
What would happen in case `waitForLatch` timed out and returned false?

Joscorbe replied:
The idea behind the Latch mechanism is to wait a few seconds until a new connection is fully established. This only happens when the connection is really fresh. After the first `clusterDescriptionChanged` is called, the variables are filled in and the latch is no longer used.

If the latch times out, the ServerAddress returned will be null, which shouldn't be a problem as it has the `Nullable` annotation and when it's used it is checked for null. For example: https://github.com/apache/jackrabbit-oak/blob/c6ddcc55bee3de915459af01e91edad32d538f3d/oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoStatusTest.java#L207-L211

Joscorbe replied:
Following your comment, I'm adding the `Nullable` annotation to the 3 methods in [MongoConnection.java](https://github.com/apache/jackrabbit-oak/blob/fada62153fe314c144ba6b8ccc9155438fe8f2a8/oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java), so it clearly states the returned object can be null:
https://github.com/apache/jackrabbit-oak/blob/fada62153fe314c144ba6b8ccc9155438fe8f2a8/oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java#L352

And the following ones.

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDBConnection.java:
```diff
@@ -61,16 +64,19 @@ final class MongoDBConnection {
-        options.addServerMonitorListener(serverMonitorListener);
-        options.socketKeepAlive(socketKeepAlive);
+        MongoClientSettings.Builder options = MongoConnection.getDefaultBuilder();
+        options.applyToServerSettings((builder) -> {
```
No need for parentheses and curly braces for one line inside the lambda.
```suggestion
        options.applyToServerSettings(builder -> builder.addServerMonitorListener(serverMonitorListener));
```

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDBConnection.java:
```diff
@@ -61,16 +64,19 @@ final class MongoDBConnection {
+        });
         if (socketTimeout > 0) {
-            options.socketTimeout(socketTimeout);
+            options.applyToSocketSettings((builder) -> {
```
Please see the above comment.

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java:
```diff
@@ -142,41 +153,53 @@ public void close() {
+        return MongoClientSettings.builder()
+                .applicationName("Oak DocumentMK")
+                .applyToConnectionPoolSettings(builder -> {
+                    builder.maxWaitTime(DEFAULT_MAX_WAIT_TIME, TimeUnit.MILLISECONDS);
```
please add a static import for TimeUnit.MILLISECONDS.

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java:
```diff
@@ -300,4 +324,61 @@ public static ReadConcernLevel readConcernLevel(ReadConcern readConcern) {
+        return null;
+    }
+
+    private static MongoClusterListener getOakClusterListener(@NotNull MongoClient client) {
```
Make the return type of this API optional and make the client use fluent coding api's.

rishabhdaim replied:
```suggestion
    private static Optional<MongoClusterListener> getOakClusterListener(@NotNull MongoClient client) {
    return client.getClusterDescription().getClusterSettings().getClusterListeners().stream().filter(clusterListener -> clusterListener instanceof MongoClusterListener).map(clusterListener -> (MongoClusterListener) clusterListener).findFirst();
```

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java:
```diff
@@ -300,4 +324,61 @@ public static ReadConcernLevel readConcernLevel(ReadConcern readConcern) {
+     * @return {@ServerAddress} of the primary. {@null} if not connected or no listener was available.
+     */
+    public static ServerAddress getPrimaryAddress(@NotNull MongoClient client) {
+        MongoClusterListener listener = getOakClusterListener(client);
```
After making `getOakClusterListener()` to return optional this can be re-written as follows:

```suggestion
        MongoClusterListener listener = getOakClusterListener(client).map(l -> l.getPrimaryAddress()).orElse(null);
```

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java:
```diff
@@ -300,4 +324,61 @@ public static ReadConcernLevel readConcernLevel(ReadConcern readConcern) {
+     * @return {@ServerAddress} of the cluster. {@null} if not connected or no listener was available.
+     */
+    public static ServerAddress getAddress(@NotNull MongoClient client) {
+        MongoClusterListener listener = getOakClusterListener(client);
```
See comment for `getPrimaryAddress()`

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoClusterListener.java:
```diff
@@ -0,0 +1,82 @@
+    private void waitForLatch() {
+        try {
+            LATCH.await(LATCH_AWAIT_TIMEOUT, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {}
```
I would add a `log.error` msg here instead of swallowing the exception.

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java:
```diff
@@ -300,4 +324,61 @@ public static ReadConcernLevel readConcernLevel(ReadConcern readConcern) {
+     * @return {@code true} if part of Replica Set, {@code false} otherwise.
+     */
+    public static boolean isReplicaSet(@NotNull MongoClient client) {
+        MongoClusterListener listener = getOakClusterListener(client);
```
See the comment for `getPrimaryAddress()`

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/DataStoreCheckCommand.java:
```diff
@@ -188,9 +188,9 @@ static int checkDataStore(String... args) {
-                if (source.startsWith(MongoURI.MONGODB_PREFIX)) {
-                    MongoClientURI uri = new MongoClientURI(source);
-                    MongoClient client = new MongoClient(uri);
+                if (source.startsWith(MongoConnection.MONGODB_PREFIX)) {
```
Implementation of `options.valueOf(store)` clearly states that it can return an null value if it returns an `emptyList()`. I think we should handle the null check here as well.

```suggestion
                if (StringUtils.startsWith(source, MongoConnection.MONGODB_PREFIX)) {
```

where `StringUtils` is from `org.apache.commons.lang3.StringUtils`

Joscorbe committed:
Addressed changes proposed in PR #295

Joscorbe commented:
Thanks @mreutegg and @rishabhdaim for your reviews. I think I have addressed all the suggestions in the code.
I would appreciate if you have more comments or give green light to merge it.

mreutegg reviewed (approved):
Looks good to me.

Joscorbe committed:
Merge remote-tracking branch 'oak/trunk' into mongo-4.2

Joscorbe committed:
Bump Mongo Java Driver version to 4.7.2

Joscorbe committed:
Removed unused Mongo dependency that requires Mongo Java Driver 2.10

Joscorbe committed:
Updated newly added MongoDBConfigTest class to Mongo 4.7 API.

Joscorbe committed:
Merge branch 'apache:trunk' into mongo-4.2

Joscorbe committed:
OAK-9447: Bump Mongo Driver version to 4.8.0
